third_party/amd/lib/TritonAMDGPUTransforms/AccelerateAMDMatmul.cpp

```diff
@@ -52,20 +96,21 @@ warpsPerTile(Operation *dotOp, ArrayRef<int64_t> shape, int numWarps,
   if (rank == 3)
     return {(unsigned)numWarps, 1, 1};
 
-  auto filter = [dotOp](Operation *op) {
-    return op->getParentRegion() == dotOp->getParentRegion();
-  };
-  ForwardSliceOptions fwdOpt;
-  fwdOpt.filter = filter;
-  BackwardSliceOptions bwdOpt;
-  bwdOpt.omitBlockArguments = true;
-  bwdOpt.filter = filter;
-  auto slices = getSlice(dotOp, bwdOpt, fwdOpt);
-  for (Operation *op : slices) {
-    if (isa<mlir::triton::DotOpInterface>(op) && (op != dotOp))
+  // For FA-like pattern, i.e. result of 1st tl.dot is used as the
+  // operand(0) of the 2nd dot.
+  // We use {numWaprs, 1} for both tl.dots
+  // However, if shape[0] is small to hold one warp size, all other
+  // warps are holding redundant values. In this case, we use
+  // {1, numWarps} for the 2nd tl.dot to save registers
+  auto ttDotOp = dyn_cast<tt::DotOp>(dotOp);
+  if (isChainDotHead(ttDotOp) || isChainDotTail(ttDotOp)) {
+    if ((shape[0] == shapePerWarp.first) && isChainDotTail(ttDotOp))
```

Comment:
You can use a local variable to save the result for `isChainDotTail` to avoid compute it again. (The C++ compiler should do it but I'm not 100% sure.) Also this is a specific case. I suspect we want to swap here as long as it's more beneficial to distribute along second dot's N dim? That is, `ceildiv(shape[0], shapePerWarp.first) < ceildiv(shape[1], shapePerWarp.second)`?